### ISSUE TITLE
Dot products over structured input with multivariate outputs

### ISSUE THREAD

**Issue by patwa67** [rstudio/keras#694]
Hi,
I have a model with two output variables (quant and bin) and want to do p (in the example 9723) number of dot products over the structured weights (between the [1:3] parts from quant_pred and bin_pred), but I cannot figure out how to implement it. A short version of the code is:
```
# One-hot encoding input
X_train <- to_categorical(xtrain, 3)
X_test <- to_categorical(xtest, 3)

> str(X_train)
 num [1:2326, 1:9723, 1:3] 0 0 0 0 0 0 0 0 0 0 ...

#Reshape data into Kearas format
phot<-dim(X_train)[2]*3
X_train <- array_reshape(X_train, c(nrow(X_train), phot))
X_test <- array_reshape(X_test, c(nrow(X_test), phot))

#The model starts here
input <- layer_input(shape = dim(X_train)[2])

quant_pred <- input %>%
  layer_dense(units = 1, activation='linear',name = "quant")
  mod_quant<-callback_model_checkpoint("modquant_{epoch:04d}.h5",
                                   verbose=0,save_weights_only = T)

bin_pred <- input %>%
  layer_dense(units = 1,activation='sigmoid', name = "bin")
  mod_bin<-callback_model_checkpoint("modbin_{epoch:04d}.h5",
                                     verbose=0,save_weights_only = T)

model <- keras_model(
  input,list(quant_pred, bin_pred)
)
#and ends here
```

**Comment by skeydan** [comment]
You can extract the weights from both layers using `get_weights`:

https://tensorflow.rstudio.com/keras/reference/get_weights.html

**Comment by patwa67** [comment]
Well, I'm not sure that is what I want. My goal is that the dot product (correlation) should provide information to the model in the training.

**Comment by skeydan** [comment]
Then probably instead of

> number of dot products over the structured weights

you want to compute dot products of intermediate tensors?

In that case you can use `layer_dot` to combine 2 layers, or in case you need to do further manipulation/subselecting, put that into a `layer_lambda` (or even a custom layer, see https://tensorflow.rstudio.com/keras/articles/custom_layers.html), like so:

```
library(keras)

input1 <- layer_input(shape = 10)
input2 <- layer_input(shape = 10)

dense1 <- input1 %>% layer_dense(units = 17)
dense2 <- input2 %>% layer_dense(units = 17)

### simple dot product
#dot <- layer_dot(list(dense1, dense2), axes = 1)
### custom layer that could be enhanced to do more than computing the dot product
dot <- layer_lambda(list(dense1, dense2),
                    f = function(x) {
                      k_batch_dot(x[[1]], x[[2]], axes = 2)
                    })


model <- keras_model(list(input1, input2), dot)

x1 <- matrix(rep(1:10, 29), ncol = 10, byrow = TRUE)
x2 <- x1
y <- (1:10)^2 %>% sum() %>% rep(29) %>% matrix()

model %>% compile(optimizer = optimizer_adam(), loss = "mse")
model %>% fit(list(x1, x2), y, epochs = 100)

model %>% predict(list(x1, x2))

```

**Comment by patwa67** [comment]
OK, thanks, this is more in line with what I'm looking for, but I still have some issues. Note that I have two output variables. I don't want the dot layer to be the output layer, just an intermediate trainable layer. However, when I try the code below the dot layer is not included in the model. How do I include it, and how do I get the values from it? And how do I exclude the bias from the dot product operation?
```
library(keras)

input <- layer_input(shape = 10)

dense1 <- input %>% layer_dense(units = 1)
dense2 <- input %>% layer_dense(units = 1)

### simple dot product
#dot <- layer_dot(list(dense1, dense2), axes = 1,trainable =TRUE)
### custom layer that could be enhanced to do more than computing the dot product
dot <- layer_lambda(list(dense1, dense2),
                    f = function(x) {
                      k_batch_dot(x[[1]], x[[2]], axes = 2)
                    })

model <- keras_model(input,list(dense1, dense2))

summary(model)

x <- matrix(rep(1:10, 29), ncol = 10, byrow = TRUE)+rnorm(29*10)

y1 <- (1:10) %>% sum() %>% rep(29) %>% matrix()
y2 <- y1+rnorm(29)
 
model %>% compile(optimizer = optimizer_adam(), loss = c("mse","mse"))
model %>% fit(x, list(y1,y2), epochs = 100)
```

**Comment by skeydan** [comment]
This would include another layer after the `dot` layer:

```
input <- layer_input(shape = 10)

dense1 <- input %>% layer_dense(units = 1)
dense2 <- input %>% layer_dense(units = 1)

### simple dot product
#dot <- layer_dot(list(dense1, dense2), axes = 1,trainable =TRUE)
### custom layer that could be enhanced to do more than computing the dot product
dot <- layer_lambda(list(dense1, dense2),
                    f = function(x) {
                      k_batch_dot(x[[1]], x[[2]], axes = 2)
                    })

dense3 <- dot %>% layer_dense(units = 1)
model <- keras_model(input, dense3)

summary(model)

x <- matrix(rep(1:10, 29), ncol = 10, byrow = TRUE)+rnorm(29*10)

y <- (1:10) %>% sum() %>% rep(29) %>% matrix()

model %>% compile(optimizer = optimizer_adam(), loss = "mse")
model %>% fit(x, y, epochs = 100)
```

> And how do I exclude the bias from the dot product operation?

`bias` is part of the weights, not the activations, so I don't see the connection...
In general for how to have an e.g. dense layer without bias, see https://tensorflow.rstudio.com/keras/reference/layer_dense.html

**Comment by patwa67** [comment]
No, now the dot layer is connected to the output. I want  dense1 and dense2 to be connected to the output, or if you look at my first model quant_pred and bin_pred. To put a loss function over the dot product makes no sense. I want separate loss functions over dense1 and dense2, and it is the dot layer that should connect the these two separate models.

**Comment by skeydan** [comment]
I don't understand what you want to do. How about you construct a model, based on my last example or the above, so that even if that produces an error I can see what it's supposed to do?

**Comment by patwa67** [comment]
The model is above. The problem is that the dot product layer isn't included in the model.

**Comment by skeydan** [comment]
Well, I'm not a mind reader... somehow you need to figure out what kind of graph you want the model to be.

You might get some inspiration here: https://tensorflow.rstudio.com/keras/articles/functional_api.html
Meaning, the functional API offers a lot of options for graph construction, some of that may apply to your requirements.

Your using the word "correlation" above made me think you might be interested in this, too: https://blogs.rstudio.com/tensorflow/posts/2018-09-26-embeddings-recommender/
